### PR TITLE
Added error types for each case for the status code

### PR DESCRIPTION
**Pull request by majioa**

### CODE REVIEW

File: lib/storekit/error.rb [hunk 1]
```diff
@@ -25,5 +25,41 @@ def initialize(status_code)
     def server_error?
       [21004, 21007, 21008].include?(status_code) || message == 'Unknown error'
     end
+
+    def self.new status_code
+      return super if self != ValidationError
+
+      @status_code = status_code
+      type =
+      case status_code
```
Comment (dteoh) [inline lib/storekit/error.rb]: Can you follow the indentation style and case statement style in the `initialize` method?

File: lib/storekit/error.rb [hunk 2]
```diff
@@ -25,5 +25,41 @@ def initialize(status_code)
     def server_error?
       [21004, 21007, 21008].include?(status_code) || message == 'Unknown error'
     end
+
+    def self.new status_code
+      return super if self != ValidationError
+
+      @status_code = status_code
```
Comment (dteoh) [inline lib/storekit/error.rb]: This is not needed because it is being set in `initialize`.

File: lib/storekit/error.rb [hunk 3]
```diff
@@ -25,5 +25,41 @@ def initialize(status_code)
...
+        ReceiptServerUnavailableError
+      when 21006
+        SubscriptionExpiredError
+      when 21007
+        TestEnvironmentRequiredError
+      when 21008
+        ProductionEnvironmentRequiredError
+      end
+
+      type.is_a?(NilClass) && super || type.new(status_code)
```
Comment (dteoh) [inline lib/storekit/error.rb]: Add some new unit tests for this and at least one of the error subclass branches in the case statement.

File: lib/storekit/error.rb [hunk 4]
```diff
@@ -25,5 +25,41 @@ def initialize(status_code)
...
+    end
   end
+
+  class InvalidJsonError < ValidationError; end
+  class InvalidReceiptFormatError < ValidationError; end
+  class ReceiptAuthenticationError < ValidationError; end
+  class SharedSecretInvalidError < ValidationError; end
```
Comment (dteoh) [inline lib/storekit/error.rb]: Since the first two new error classes have the `Invalid` prefix, let's do the same for this one as well.

**Review by dteoh — CHANGES_REQUESTED** [review]
Thank you for your PR. I have left you some comments.

**Comment by majioa** [comment]
Fixed some.

**Review by dteoh — APPROVED** [review]

**Comment by dteoh** [comment]
Thanks

**Pull request merged**